### PR TITLE
feat: preemptive

### PR DESCRIPTION
**Pull request by kobayu858**
## Description

- Preemptive implementation
- Writing to logs will be implemented after this PR

## Related links

None.

## Pre-review checklist for the PR author

- [x] I've confirmed the [self-checklist](https://docs.google.com/spreadsheets/d/1p5KMcm752iCySHJz0xL1qf6fcLDKprbo/edit#gid=863618628).
- [x] The PR is ready for review.

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

### CODE REVIEW

File: lib/src/dag_set_scheduler.rs [hunk 1]
```diff
@@ -182,7 +187,7 @@ pub trait DAGSetSchedulerBase<T: ProcessorBase + Clone> {
         log.calculate_response_time();
     }
 
-    fn schedule(&mut self) -> i32 {
+    fn schedule(&mut self, preemptive_key: PreemptiveType) -> i32 {
```
Comment (atsushi421) [inline lib/src/dag_set_scheduler.rs:190]:
```suggestion
    fn schedule(&mut self, preemptive_type: PreemptiveType) -> i32 {
```

File: lib/src/dag_set_scheduler.rs [hunk 2]
```diff
@@ -196,20 +201,49 @@ pub trait DAGSetSchedulerBase<T: ProcessorBase + Clone> {
                 });
             }
 
-            // Allocate the nodes of ready_queue to idle cores
+            // Allocate nodes as long as there are idle cores, and attempt to preempt when all cores are busy.
```
Comment (atsushi421) [inline lib/src/dag_set_scheduler.rs:204]: The following implementations are clearer:

```
            // Allocate nodes as long as there are idle cores, and attempt to preempt when all cores are busy.
            while !ready_queue.is_empty() {
                let processor = self.get_processor_mut();

                if let Some(idle_core_index) = processor.get_idle_core_index() {
                    // Allocate the node to the idle core
                    let node_data = ready_queue.pop_first().unwrap().convert_node_data();
                    self.allocate_node(
                        &node_data,
                        idle_core_index,
                        managers[node_data.get_params_value("dag_id") as usize].get_release_count()
                            as usize,
                    );
                } else if let PreemptiveType::Preemptive {
                    key: preemptive_key,
                } = &preemptive_key
                {
                    // If all cores are busy and a preemptive_key exists, attempt preemption
                    let (max_value, core_index) =
                        processor.get_max_value_and_index(preemptive_key).unwrap();
                    let ready_node_data_value = ready_queue
                        .first()
                        .unwrap()
                        .convert_node_data()
                        .get_params_value(preemptive_key);

                    if max_value > ready_node_data_value {
                        // Preempt the node with the lowest priority
                        let suspended_node_data = processor.suspend_execution(core_index).unwrap();
                        processor.allocate_specific_core(
                            core_index,
                            &ready_queue.pop_first().unwrap().convert_node_data(),
                        );
                        ready_queue.insert(NodeDataWrapper {
                            node_data: suspended_node_data,
                        });
                    }
                } else {
                    break; // No core is idle and no preemption possible, exit the loop.
                }
            }
```

**Review by atsushi421 — CHANGES_REQUESTED** [review]

File: lib/src/dag_set_scheduler.rs (hunk 2 above)
Comment (kobayu858) [reply lib/src/dag_set_scheduler.rs:204]: Thanks for the refactoring.

I added a break to exit the loop when no preemptive happens with all cores busy and the preemptive type specified.

**Review by atsushi421 — APPROVED** [review]
LGTM

**Pull request merged**